be/src/storage/lake/vacuum.cpp

```diff
@@ -631,4 +633,232 @@ void delete_txn_log(TabletManager* tablet_mgr, const DeleteTxnLogRequest& reques
+        }
+        if (txn_log->has_op_schema_change()) {
+            for (const auto& rowset : txn_log->op_schema_change().rowsets()) {
+                check_rowset(rowset);
+            }
+            if (txn_log->op_schema_change().has_delvec_meta()) {
+                check_delvecs(txn_log->op_schema_change().delvec_meta());
+            }
+        }
+    }
+
+    LOG(INFO) << "Found " << datafiles.size() << " orphan files";
+    return datafiles;
+}
+
+// root_location is a partition dir in s3
+Status datafile_gc(std::string_view root_location, int64_t expire_seconds) {
+    ASSIGN_OR_RETURN(auto fs, FileSystem::CreateSharedFromString(root_location));
+
+    const auto metadata_root_location = join_path(root_location, kMetadataDirectoryName);
+    const auto txn_log_root_location = join_path(root_location, kTxnLogDirectoryName);
+    const auto segment_root_location = join_path(root_location, kSegmentDirectoryName);
+
+    LOG(INFO) << "Start to list " << metadata_root_location;
+
+    std::set<std::string> tablet_metadatas;
+    RETURN_IF_ERROR(list_tablet_metadata(metadata_root_location, &tablet_metadatas));
+
+    if (tablet_metadatas.empty()) {
+        LOG(INFO) << "Skipped datafile GC of " << root_location << " because there is no tablet metadata";
+        return Status::OK();
+    }
+
+    LOG(INFO) << "Found " << tablet_metadatas.size() << " meta files";
+
+    // List txn log
+    std::vector<std::string> txn_logs;
+    /*
+    RETURN_IF_ERROR(ignore_not_found(fs->iterate_dir(txn_log_root_location, [&](std::string_view name) {
+        txn_logs.emplace_back(name);
+        return true;
+    })));
+    */
+
+    // Find orphan data files, include segment, del, and delvec
+    ASSIGN_OR_RETURN(auto orphan_datafiles,
+                     find_orphan_datafiles(root_location, tablet_metadatas, txn_logs, expire_seconds));
+
+    LOG(INFO) << "Found " << orphan_datafiles.size() << " orphan data files:";
+
+    std::vector<std::string> files_to_delete;
+    int64_t bytes_to_delete = 0;
+    for (const auto& [file, entry] : orphan_datafiles) {
+        files_to_delete.push_back(join_path(segment_root_location, file));
+        bytes_to_delete += entry.size.value_or(0);
+        auto time = entry.mtime.value_or(0);
+        LOG(INFO) << "file: " << file << ", size: " << entry.size.value_or(0)
+                  << ", time: " << std::put_time(std::localtime(&time), "%Y-%m-%d %H:%M:%S");
+    }
+
+    LOG(INFO) << "Found " << orphan_datafiles.size() << " orphan data files, total size: " << bytes_to_delete;
+
+    return Status::OK();
+    /*
+    std::cout << root_location << " has " << orphan_datafiles.size()
+              << " orphan data files, total size: " << bytes_to_delete << ", delete them ? (y/n)" << std::endl;
+
+    std::string input;
+    if (std::cin >> input && input != "y") {
+        return Status::OK();
+    }
+
+    return do_delete_files(fs.get(), files_to_delete);
```

Comment:
Move to `trash` dir may be more safe.

Reply:
I want, but s3 even not support rename...